### PR TITLE
fix(events): handle errors and undefined in payload [INFRA-827] [INFRA-828]

### PR DESCRIPTION
**Pull request by kschelonka**
## Goal
Handle errors and undefined save data during analytics event emission.

- Log Sentry warning event if save is null (can happen due to async promise resolution e.g. if a save is deleted)
- Catch all other downstream issues when creating tag payload and send exception to Sentry with warning level
- Use tag model to resolve tags
- Simplify the event emitter logic: don't need to accept Promise<SavedItem> anymore, just SavedItem

Jira tickets: [INFRA-827], [INFRA-828]

[INFRA-827]: https://getpocket.atlassian.net/browse/INFRA-827?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[INFRA-828]: https://getpocket.atlassian.net/browse/INFRA-828?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/server/context.spec.ts [hunk 1]
```diff
@@ -19,74 +22,135 @@ describe('context', () => {
...
+        request: {
+          headers: { userid: '1', apiid: '0' },
+        },
+        dbClient: jest.fn() as unknown as Knex,
+        eventEmitter: null,
+      });
+    });
 
-  afterEach(() => jest.clearAllMocks());
+    afterEach(() => jest.clearAllMocks());
```
Comment (kschelonka) [inline src/server/context.spec.ts:105]: this was just indented/brought into a describe block

File: src/server/context.spec.ts [hunk 2]
```diff
@@ -19,74 +22,135 @@ describe('context', () => {
...
+    beforeEach(() => {
+      sinon.restore();
+      sentryEventSpy = sinon.spy(Sentry, 'captureEvent');
+      sentryExceptionSpy = sinon.spy(Sentry, 'captureException');
+    });
+    afterAll(() => sinon.restore());
+    it('should log a warning to Sentry if save is undefined', async () => {
+      await context.emitItemEvent(EventType.ARCHIVE_ITEM, undefined);
+      expect(sentryEventSpy.callCount).to.equal(1);
+      const event = sentryEventSpy.getCall(0).args[0];
```
Comment (Herraj) [inline src/server/context.spec.ts:44]: is `getCall(0).args[0]` just fetching us the arguments used to make the first call with this spy object?

File: src/server/context.spec.ts [hunk 3]
```diff
@@ -19,74 +22,135 @@ describe('context', () => {
...
+      expect(batchUrlFnSpy.mock.calls[0][0]).to.deep.equal(['dont-care.com']);
+      expect(savedItems).to.deep.equal(savedItem);
+    });
+    it('Uses the same dataloader for every load request', async () => {
+      // Referencing the loader 2x should return the same object
+      const loader = context.dataLoaders.savedItemsByUrl;
+      const loaderAgain = context.dataLoaders.savedItemsByUrl;
+      await loader.load('dont-care.com');
+      // At this point both loaders should have filled cache since referencing same object
+      expect(Array.from((loader as any)._cacheMap.keys())).to.contain(
```
Comment (Herraj) [inline src/server/context.spec.ts:123]: `as any` to satisfy typescript?

File: src/server/context.ts [hunk 4]
```diff
@@ -106,15 +106,28 @@ export class ContextManager implements IContext {
...
+    if (savedItem == null) {
+      Sentry.captureEvent({
+        message: 'Save was null or undefined when generating event payload',
+        level: 'warning',
+      });
+      return;
+    }
+    try {
+      const tags = (await this.models.tag.getBySaveId(savedItem.id)).map(
+        (_) => _.name
```
Comment (Herraj) [inline src/server/context.ts:123]: Wondering why not use a descriptive variable name instead of `_` 🤔

**Review by Herraj — APPROVED** [review]

File: src/server/context.spec.ts (hunk 3 above)
Comment (kschelonka) [reply src/server/context.spec.ts:123]: yeah I think so. this was just idnented

File: src/server/context.spec.ts (hunk 2 above)
Comment (kschelonka) [reply src/server/context.spec.ts:44]: yep!

File: src/server/context.ts (hunk 4 above)
Comment (kschelonka) [reply src/server/context.ts:123]: making a note to remove scala idioms in the future

**Pull request merged**